### DISCUSSION TITLE
parse vs accept

### DISCUSSION THREAD

**Discussion by bajaj689** [nlohmann/json#3522] [Q&A] [answered]
Hi All, I am writing an API where the request will contain a JSON string.
After reading this request, I want to check if JSON is valid or not. For that I came across the function named "accept". I want to know what is the use of the function "parse", because this function can also tell us whether JSON string was valid or not, ofcourse it has ability to throw expections a/c to parameter passed. 
My question is  that usage wise,  when should parse be used and where should accept be used ? Do they provide same functionality?

**Comment by nlohmann** [comment, marked as answer]
`accept` is to check validity, so you only get a boolean as result.

`parse` is to get a `basic_json` object from the input.

**Reply by bajaj689** [reply]
Thanks a lot for the answer. Can there be a case where accept function says JSON is valid, but the subsequent parse function fails for the same input.
Does parse function "check validity plus deserialize" ?

**Reply by nlohmann** [reply]
No. If `accept` returns `true`, `parse` does not throw an exception. Any other behavior would be a bug.

**Reply by bajaj689** [reply]
![image](https://user-images.githubusercontent.com/30772186/172575815-9472eedb-f2f7-4c18-8b08-d2dcd95f8fe7.png)
![image](https://user-images.githubusercontent.com/30772186/172575887-78b2a28d-1f4b-4bac-a179-c95fd5d359dc.png)


In the first image, the functional signature shows 4 arguments. In image 2, it rightly says that 3rd argument is for specifying exception flag. But If I give 4 arguments, it gives compilation error._"no matching function for call to ‘nlohmann::basic_json<>::parse(char [1024], std::nullptr_t, bool, bool, "_ 

Does the signature need to be updated in document? Or I am getting this wrong ?

**Reply by falbrechtskirchinger** [reply]
What version are you using?

**Reply by bajaj689** [reply]
2.1.1

**Reply by falbrechtskirchinger** [reply]
That is ancient. 3.11.0 will be released soon. The latest available version is 3.10.5, I believe.

**Reply by bajaj689** [reply]
Ok, Thanks

**Comment by lefebvresam** [comment]
If you use a file, for an unknown reason, the file variabele can  not be used twice after each other in parse or accept. The second time an empty file is reading.

```
FILE *file = fopen(fullpath.c_str(), "rb");
if (!json::accept(file)) {
    unmountSD();
    ERROR("invalid json in %s", fullpath.c_str());
    return 1;
}
try{
   json app = json::parse(file); // this throws error 101 unless you comment the accept line
} catch (json::parse_error& ex) {
    unmountSD();
    ERROR("%s", ex.what());
    return 1;
}
```

[json.exception.parse_error.101] parse error at line 1, column 1: attempting to parse an empty input; check that your input string or stream contains the expected JSON

**Reply by lefebvresam** [reply]
Solution to avoid try/catch to save flash memory, which otherwise gives 5 pages extra code:

```
json app = json::parse(file, nullptr, false);
if (app.is_discarded()) {
    unmountSD();
    WARNING("invalid json in %s", fullpath.c_str());
    return 0;
}
```

**Reply by nlohmann** [reply]
That is expected behavior.

Reading from a C FILE* or a C++ std::ifstream twice without proper handling doesn't work as expected because file streams maintain an internal position marker that advances with each read operation.

Before reading a second time, you need to rewind the file stream.

**Reply by lefebvresam** [reply]
Any example?

**Reply by natri0** [reply]
If you're reading from a `FILE *`:
```cpp
FILE *fp = fopen(/* ... */);
if (json::accept(file)) {
  fseek(fp, 0, SEEK_SET);
  auto j = json::parse(fp);
  printf("read json: %s", j.dump().c_str());
} else {
  fprintf(stderr, "not valid json\n");
}
```

**Reply by nlohmann** [reply]
Why don't you just parse the JSON immediately? This will take twice as long as you need to go through the file twice.

**Reply by lefebvresam** [reply]
Ok now I use the method with discarded and this works.
